datafusion/functions/src/datetime/date_bin.rs

```diff
@@ -146,7 +162,44 @@ impl DateBinFunc {
                     DataType::Interval(DayTime),
                     Timestamp(array_type, Some(TIMEZONE_WILDCARD.into())),
                 ]),
-            ]
+            ];
+
+            match array_type {
+                Second | Millisecond => {
+                    v.append(&mut vec![
+                        Exact(vec![
+                            DataType::Interval(MonthDayNano),
+                            Time32(array_type),
+                            Time32(array_type),
```

Comment:
Could you explain why for time we require the expression & origin arguments to have the same granularity in terms of time units, whereas for timestamps above we seem to always have origin as Nanoseconds regardless of the expression granularity?